File: core/rust/grain-core/src/ledger.rs
use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::{Diag, GrainError, GrainResult};

#[derive(Debug, Clone, Deserialize)]
pub struct LedgerEvent {
    pub t: String,
    pub ak: String,
    pub seq: u64,
    pub payload_cid: String,
    pub body: Value,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct LedgerTotals {
    pub sum_mean: BTreeMap<String, i64>,
    pub sum_var: BTreeMap<String, i64>,
    pub diag_contains: Vec<String>,
}

pub fn reduce_ledger(root_kid: &str, events: &[LedgerEvent]) -> GrainResult<LedgerTotals> {
    let mut diagnostics: BTreeSet<Diag> = BTreeSet::new();

    let mut grants: BTreeSet<String> = BTreeSet::new();
    let mut revokes: BTreeSet<String> = BTreeSet::new();

    for ev in events {
        match ev.t.as_str() {
            "DeviceKeyGrant" => {
                if ev.ak == root_kid {
                    if let Some(grant_ak) = ev.body.get("grant_ak").and_then(Value::as_str) {
                        grants.insert(grant_ak.to_string());
                    }
                } else {
                    diagnostics.insert(Diag::UnauthorizedGrantIgnored);
                }
            }
            "DeviceKeyRevoke" => {
                if ev.ak == root_kid {
                    if let Some(revoke_ak) = ev.body.get("revoke_ak").and_then(Value::as_str) {
                        revokes.insert(revoke_ak.to_string());
                    }
                } else {
                    diagnostics.insert(Diag::UnauthorizedGrantIgnored);
                }
            }
            _ => {}
        }
    }

    let is_authorized = |ak: &str| -> bool {
        if ak == root_kid {
            return true;
        }
        grants.contains(ak) && !revokes.contains(ak)
    };

    let mut authorized_events: Vec<&LedgerEvent> = Vec::new();
    for ev in events {
        if !is_authorized(&ev.ak) {
            if revokes.contains(&ev.ak) {
                diagnostics.insert(Diag::AkRevoked);
            }
            continue;
        }
        authorized_events.push(ev);
    }

    let mut pair_payloads: BTreeMap<(String, u64), BTreeSet<String>> = BTreeMap::new();
    for ev in &authorized_events {
        pair_payloads
            .entry((ev.ak.clone(), ev.seq))
            .or_default()
            .insert(ev.payload_cid.clone());
    }

    let conflicted: BTreeSet<(String, u64)> = pair_payloads
        .iter()
        .filter_map(|(k, payloads)| if payloads.len() > 1 { Some(k.clone()) } else { None })
        .collect();

    if !conflicted.is_empty() {
        diagnostics.insert(Diag::SeqConflict);
    }

    let mut sum_mean: i128 = 0;
    let mut sum_var: i128 = 0;

    let mut seen_exact: BTreeSet<(String, u64, String)> = BTreeSet::new();

    for ev in authorized_events {
        if conflicted.contains(&(ev.ak.clone(), ev.seq)) {
            continue;
        }
        if !seen_exact.insert((ev.ak.clone(), ev.seq, ev.payload_cid.clone())) {
            // set-union semantics: exact duplicates are idempotent
            continue;
        }
        if ev.t != "IntakeEvent" {
            continue;
        }

        let mean_kcal = ev
            .body
            .get("mean")
            .and_then(|m| m.get("kcal"))
            .and_then(Value::as_i64)
            .ok_or_else(|| GrainError::from_diag(Diag::Schema))?;

        let var_kcal = ev
            .body
            .get("var")
            .and_then(|m| m.get("kcal"))
            .and_then(Value::as_i64)
            .ok_or_else(|| GrainError::from_diag(Diag::Schema))?;

        if var_kcal < 0 {
            return Err(GrainError::from_diag(Diag::Schema));
        }

        sum_mean += mean_kcal as i128;
        sum_var += var_kcal as i128;
    }

    if sum_mean > i64::MAX as i128 || sum_mean < i64::MIN as i128 {
        return Err(GrainError::from_diag(Diag::Overflow));
    }
    if sum_var > i64::MAX as i128 || sum_var < 0 {
        return Err(GrainError::from_diag(Diag::Overflow));
    }

    let mut out_mean = BTreeMap::new();
    out_mean.insert("kcal".to_string(), sum_mean as i64);

    let mut out_var = BTreeMap::new();
    out_var.insert("kcal".to_string(), sum_var as i64);

    let diag_contains = diagnostics.into_iter().map(|d| d.code().to_string()).collect();

    Ok(LedgerTotals {
        sum_mean: out_mean,
        sum_var: out_var,
        diag_contains,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn intake_event(ak: &str, seq: u64, payload_cid: &str, mean: i64, var: i64) -> LedgerEvent {
        LedgerEvent {
            t: "IntakeEvent".to_string(),
            ak: ak.to_string(),
            seq,
            payload_cid: payload_cid.to_string(),
            body: json!({
                "mean": { "kcal": mean },
                "var": { "kcal": var },
            }),
        }
    }

    #[test]
    fn exact_duplicate_intake_event_is_idempotent() {
        let event = intake_event("root", 1, "cid-1", 7, 3);
        let totals = reduce_ledger("root", &[event.clone(), event]).unwrap();

        assert_eq!(totals.sum_mean.get("kcal"), Some(&7));
        assert_eq!(totals.sum_var.get("kcal"), Some(&3));
        assert!(totals.diag_contains.is_empty());
    }

    #[test]
    fn conflicting_sequence_is_dropped_and_flagged() {
        let totals = reduce_ledger(
            "root",
            &[
                intake_event("root", 1, "cid-a", 7, 3),
                intake_event("root", 1, "cid-b", 9, 5),
            ],
        )
        .unwrap();

        assert_eq!(totals.sum_mean.get("kcal"), Some(&0));
        assert_eq!(totals.sum_var.get("kcal"), Some(&0));
        assert!(totals.diag_contains.contains(&Diag::SeqConflict.code().to_string()));
    }

    #[test]
    fn unauthorized_grant_is_ignored_and_authorized_events_still_reduce() {
        let totals = reduce_ledger(
            "root",
            &[
                LedgerEvent {
                    t: "DeviceKeyGrant".to_string(),
                    ak: "intruder".to_string(),
                    seq: 1,
                    payload_cid: "grant-cid".to_string(),
                    body: json!({ "grant_ak": "kid-1" }),
                },
                intake_event("root", 2, "cid-1", 7, 3),
            ],
        )
        .unwrap();

        assert_eq!(totals.sum_mean.get("kcal"), Some(&7));
        assert_eq!(totals.sum_var.get("kcal"), Some(&3));
        assert!(totals.diag_contains.contains(&Diag::UnauthorizedGrantIgnored.code().to_string()));
    }

    #[test]
    fn revoked_authorized_events_are_flagged_and_dropped() {
        let totals = reduce_ledger(
            "root",
            &[
                LedgerEvent {
                    t: "DeviceKeyGrant".to_string(),
                    ak: "root".to_string(),
                    seq: 1,
                    payload_cid: "grant-cid".to_string(),
                    body: json!({ "grant_ak": "kid-1" }),
                },
                LedgerEvent {
                    t: "DeviceKeyRevoke".to_string(),
                    ak: "root".to_string(),
                    seq: 2,
                    payload_cid: "revoke-cid".to_string(),
                    body: json!({ "revoke_ak": "kid-1" }),
                },
                intake_event("kid-1", 3, "cid-1", 7, 3),
            ],
        )
        .unwrap();

        assert_eq!(totals.sum_mean.get("kcal"), Some(&0));
        assert_eq!(totals.sum_var.get("kcal"), Some(&0));
        assert!(totals.diag_contains.contains(&Diag::AkRevoked.code().to_string()));
    }
}

File: runner/typescript/src/ops/ledger.ts
import { GrainDiagError } from "../types.js";
import type { Json, OperationActual } from "../types.js";
import { normalizeDiag } from "../utils.js";
import { parseInteger, textFromObjectField, toObject, toText } from "./helpers.js";

const I64_MIN = -(1n << 63n);
const I64_MAX = (1n << 63n) - 1n;

type LedgerEvent = {
  t: string;
  ak: string;
  seq: bigint;
  payloadCid: string;
  body: Record<string, Json>;
};

export function opLedgerReduce(input: Record<string, Json>): OperationActual {
  const rootKid = toText(input.root_kid, "GRAIN_ERR_SCHEMA");
  const events = parseLedgerEvents(input.events);

  const diagnostics = new Set<string>();
  const grants = new Set<string>();
  const revokes = new Set<string>();

  for (const ev of events) {
    if (ev.t === "DeviceKeyGrant") {
      if (ev.ak === rootKid) {
        const grantAk = textFromObjectField(ev.body, "grant_ak");
        if (grantAk) {
          grants.add(grantAk);
        }
      } else {
        diagnostics.add("UNAUTHORIZED_GRANT_IGNORED");
      }
    }

    if (ev.t === "DeviceKeyRevoke") {
      if (ev.ak === rootKid) {
        const revokeAk = textFromObjectField(ev.body, "revoke_ak");
        if (revokeAk) {
          revokes.add(revokeAk);
        }
      } else {
        diagnostics.add("UNAUTHORIZED_GRANT_IGNORED");
      }
    }
  }

  const isAuthorized = (ak: string): boolean => {
    if (ak === rootKid) {
      return true;
    }
    return grants.has(ak) && !revokes.has(ak);
  };

  const authorizedEvents: LedgerEvent[] = [];
  for (const ev of events) {
    if (!isAuthorized(ev.ak)) {
      if (revokes.has(ev.ak)) {
        diagnostics.add("AK_REVOKED");
      }
      continue;
    }
    authorizedEvents.push(ev);
  }

  const conflicted = collectConflictedSequences(authorizedEvents);
  if (conflicted.size > 0) {
    diagnostics.add("SEQ_CONFLICT");
  }

  let sumMean = 0n;
  let sumVar = 0n;
  const seenExact = new Set<string>();

  for (const ev of authorizedEvents) {
    const pairKey = sequenceKey(ev.ak, ev.seq);
    if (conflicted.has(pairKey)) {
      continue;
    }

    const exactKey = `${pairKey}\u0000${ev.payloadCid}`;
    if (seenExact.has(exactKey)) {
      continue;
    }
    seenExact.add(exactKey);

    if (ev.t !== "IntakeEvent") {
      continue;
    }

    const meanKcal = parseBodyI64(ev.body, ["mean", "kcal"]);
    const varKcal = parseBodyI64(ev.body, ["var", "kcal"]);
    if (varKcal < 0n) {
      throw new GrainDiagError("GRAIN_ERR_SCHEMA");
    }

    sumMean += meanKcal;
    sumVar += varKcal;
  }

  if (sumMean > I64_MAX || sumMean < I64_MIN) {
    throw new GrainDiagError("GRAIN_ERR_OVERFLOW");
  }
  if (sumVar > I64_MAX || sumVar < 0n) {
    throw new GrainDiagError("GRAIN_ERR_OVERFLOW");
  }

  const out: Record<string, Json> = {
    sum_mean: { kcal: Number(sumMean) },
    sum_var: { kcal: Number(sumVar) }
  };

  const diagContains = normalizeDiag([...diagnostics]);
  if (diagContains.length > 0) {
    out.diag_contains = diagContains;
  }

  return {
    accepted: true,
    diag: [],
    out
  };
}

function collectConflictedSequences(events: LedgerEvent[]): Set<string> {
  const pairPayloads = new Map<string, Set<string>>();
  for (const ev of events) {
    const key = sequenceKey(ev.ak, ev.seq);
    if (!pairPayloads.has(key)) {
      pairPayloads.set(key, new Set());
    }
    pairPayloads.get(key)?.add(ev.payloadCid);
  }

  const conflicted = new Set<string>();
  for (const [pairKey, payloads] of pairPayloads.entries()) {
    if (payloads.size > 1) {
      conflicted.add(pairKey);
    }
  }
  return conflicted;
}

function sequenceKey(ak: string, seq: bigint): string {
  return `${ak}\u0000${seq.toString()}`;
}

function parseLedgerEvents(value: Json | undefined): LedgerEvent[] {
  if (!Array.isArray(value)) {
    throw new GrainDiagError("GRAIN_ERR_SCHEMA");
  }

  const out: LedgerEvent[] = [];
  for (const item of value) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      throw new GrainDiagError("GRAIN_ERR_SCHEMA");
    }

    const obj = item as Record<string, Json>;
    const t = toText(obj.t, "GRAIN_ERR_SCHEMA");
    const ak = toText(obj.ak, "GRAIN_ERR_SCHEMA");
    const payloadCid = toText(obj.payload_cid, "GRAIN_ERR_SCHEMA");
    const seq = parseInteger(obj.seq, "GRAIN_ERR_SCHEMA");
    if (seq < 0n) {
      throw new GrainDiagError("GRAIN_ERR_SCHEMA");
    }

    const body = toObject(obj.body, "GRAIN_ERR_SCHEMA");
    out.push({ t, ak, seq, payloadCid, body });
  }

  return out;
}

function parseBodyI64(body: Record<string, Json>, path: [string, string]): bigint {
  const l1 = body[path[0]];
  if (l1 === null || typeof l1 !== "object" || Array.isArray(l1)) {
    throw new GrainDiagError("GRAIN_ERR_SCHEMA");
  }

  const obj = l1 as Record<string, Json>;
  const v = parseInteger(obj[path[1]], "GRAIN_ERR_SCHEMA");
  if (v < I64_MIN || v > I64_MAX) {
    throw new GrainDiagError("GRAIN_ERR_SCHEMA");
  }
  return v;
}
